### PR TITLE
Bug 1714038 - Enable pagination in Compare View

### PR DESCRIPTION
**Pull request by beatrice-acasandrei**
https://mozilla-hub.atlassian.net/browse/FXP-1636

### CODE REVIEW

File: ui/perfherder/shared/Pagination.jsx [hunk 1]
```diff
@@ -19,7 +19,7 @@ class PaginationGroup extends React.Component {
       /* The first and last pagination navigation links
          aren't working correctly (icons aren't visible)
          so they haven't been added */
-      <Pagination aria-label={`Page ${currentPage}`}>
+      <Pagination aria-label={`Page ${currentPage}`} data-testid="pagination">
```
[the remark this answers is not in the document]
Comment (beatrice-acasandrei) [reply ui/perfherder/shared/Pagination.jsx]: It's not used, thanks for noticing.

File: ui/perfherder/shared/Pagination.jsx [hunk 2]
```diff
@@ -47,6 +47,7 @@ class PaginationGroup extends React.Component {
             <PaginationLink
               className="text-info"
               onClick={() => this.navigatePage(num)}
+              data-testid={`pagination-button-${num}`}
```
[the remark this answers is not in the document]
Comment (beatrice-acasandrei) [reply ui/perfherder/shared/Pagination.jsx]: It's not used, thanks for noticing.

**Review by esanuandra — APPROVED** [review]
lgtm

**Comment by beatrice-acasandrei** [comment]
@sarah-clements Hello! Could you please take a look over this when you get the chance? If everything looks ok please merge it as well. Thanks!

File: ui/perfherder/compare/CompareTableControls.jsx [hunk 3]
```diff
@@ -27,34 +29,56 @@ export default class CompareTableControls extends React.Component {
...
+
+    if (prevState.countPages !== countPages) {
+      /* eslint-disable react/no-did-update-set-state */
+      this.setState({ totalPagesList: this.generatePages(countPages) });
+    }
     if (prevProps.compareResults !== compareResults) {
       this.updateFilteredResults();
     }
+    if (params.page && params.page !== prevParams.page) {
+      /* eslint-disable react/no-did-update-set-state */
```
Comment (sarah-clements) [inline ui/perfherder/compare/CompareTableControls.jsx:56]: If you move the `/* eslint-disable react/no-did-update-set-state */` to the top of the file then you wont' need to write it twice.

File: ui/perfherder/compare/CompareTableControls.jsx [hunk 4]
```diff
@@ -27,34 +29,56 @@ export default class CompareTableControls extends React.Component {
...
 
   componentDidMount() {
     this.updateFilteredResults();
   }
 
-  componentDidUpdate(prevProps) {
+  componentDidUpdate(prevProps, prevState) {
     const { compareResults } = this.props;
+    const { countPages } = this.state;
+    const params = parseQueryParams(this.props.location.search);
```
Comment (sarah-clements) [inline ui/perfherder/compare/CompareTableControls.jsx:45]: Can do the destructuring assignment for the location prop on line 43.

File: ui/perfherder/compare/CompareTableControls.jsx [hunk 5]
```diff
@@ -27,34 +29,56 @@ export default class CompareTableControls extends React.Component {
...
+      });
+    }
   }
 
   getDefaultFilterText = (validated) => {
     const { filter } = validated;
     return filter === undefined || filter === null ? '' : filter;
   };
 
+  getDefaultFilterText = (validated) => {
```
Comment (sarah-clements) [inline ui/perfherder/compare/CompareTableControls.jsx:68]: These function is declared twice.

**Review by sarah-clements — COMMENTED** [review]
Can you remind me why this is optimizing the frontend pagination rather than switching to api/queryset filtering for the pagination? (I know we had a chat a few months ago with alexandru but I'm not remembering the details).

**Comment by beatrice-acasandrei** [comment]
> Can you remind me why this is optimizing the frontend pagination rather than switching to api/queryset filtering for the pagination? (I know we had a chat a few months ago with alexandru but I'm not remembering the details).

The way that Compare View was implemented was by making two requests, one for the base revision and one for the new revision. The responses are compared and processed on the frontend and the tables are rendered. This is the most simple and forward solution that uses the current implementation. To switch to api/queryset filtering would need more effort and could potentially introduce bugs since there is a lot of logic going on in the frontend that needs to be moved to backend. 

A part of the lagging problem was caused by the huge number of tables that were rendered at once and this pagination feature helps with that.

**Comment by beatrice-acasandrei** [comment]
@sarah-clements Thanks for the review! I added the fixes.

**Review by sarah-clements — APPROVED** [review]
Looks good

**Pull request merged**